# The Battle of the Neighborhoods (Week 2)

## Part 5A Segmenting and Clustering Neighborhoods - Brooklyn and Manhattan

## Introduction
In this section of the capstone project, we will use the Foursquare API to explore neighborhoods in Brooklyn and Manhattan. We will use the explore function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. We will use the k-means clustering algorithm to complete this task. Finally, we will use the Folium library to visualize the neighborhoods in Brooklyn and Manhattan and their emerging clusters.

## Table of Contents
1. Download and Explore Dataset
1. Explore Neighborhoods in Brooklyn and Manhattan
1. Analyze Each Neighborhood
1. Cluster Neighborhoods and Examine Clusters

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib.pyplot as plt

# import k-means from clustering stage
from sklearn.cluster import KMeans

from sklearn.metrics import silhouette_score

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0           conda-forge
    geopy:          

## 1. Download and Explore Dataset

In [2]:
!wget -q -O 'newyork_data.json' https://ibm.box.com/shared/static/fbpwbovar7lf8p5sgddm06cgipa2rxpe.json
print('Data downloaded!')

Data downloaded!


In [3]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [4]:
neighborhoods_data = newyork_data['features']

In [5]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [6]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
NYC_Geo = pd.DataFrame(columns=column_names)

In [7]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    NYC_Geo = NYC_Geo.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [8]:
NYC_Geo.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [9]:
NYC_Geo.to_csv('NYC_GEO.csv',index=False)

In [10]:
NYC_Geo['Borough'].value_counts().to_frame()

Borough
Queens              81
Brooklyn            70
Staten Island       63
Bronx               52
Manhattan           40

In [11]:
NYC_Geo.shape

(306, 4)

In [12]:
print(NYC_Geo.Borough.unique())

['Bronx' 'Manhattan' 'Brooklyn' 'Queens' 'Staten Island']


In [13]:
NYC_Geo.isnull().sum()

Borough         0
Neighborhood    0
Latitude        0
Longitude       0
dtype: int64

### Segmenting and Clustering Neighborhoods - Brooklyn and Manhattan

In [14]:
BM_Geo = NYC_Geo.loc[(NYC_Geo['Borough'] == 'Brooklyn')|(NYC_Geo['Borough'] == 'Manhattan')]
BM_Geo = BM_Geo.reset_index(drop=True)
BM_Geo.head()

Borough Neighborhood   Latitude  Longitude
0  Manhattan  Marble Hill  40.876551 -73.910660
1   Brooklyn    Bay Ridge  40.625801 -74.030621
2   Brooklyn  Bensonhurst  40.611009 -73.995180
3   Brooklyn  Sunset Park  40.645103 -74.010316
4   Brooklyn   Greenpoint  40.730201 -73.954241

In [15]:
BM_Geo.shape

(110, 4)

In [16]:
import time
start_time = time.time()

address = 'New York City, NY'

geolocator = Nominatim(user_agent="Jupyter")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

print("--- %s seconds ---" % round((time.time() - start_time), 2))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.
--- 0.43 seconds ---


In [17]:
# Create a map of Brooklyn and Manhattan using latitude and longitude values
map_BM = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(BM_Geo['Latitude'], BM_Geo['Longitude'], BM_Geo['Borough'], BM_Geo['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_BM)  
    
map_BM

In [18]:
CLIENT_ID = 'NXMXG0ON4AOS3PTHLDN2VG5KG3WZWZCBDSCCGPLJUP2BAC2C' # Foursquare ID
CLIENT_SECRET = 'YNQNJCNCPC2MYEDBP03JYXSXZNQAFHXIEQK1ASDTWVYKMU2H' # Foursquare Secret
VERSION = '20181218' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NXMXG0ON4AOS3PTHLDN2VG5KG3WZWZCBDSCCGPLJUP2BAC2C
CLIENT_SECRET:YNQNJCNCPC2MYEDBP03JYXSXZNQAFHXIEQK1ASDTWVYKMU2H


## 2. Explore Neighborhoods in Brooklyn and Manhattan

In [19]:
def getNearbyVenues(names, latitudes, longitudes, LIMIT=200, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
BM_venues = getNearbyVenues(names=BM_Geo['Neighborhood'],
                                  latitudes=BM_Geo['Latitude'],
                                  longitudes=BM_Geo['Longitude'],
                                  LIMIT=200)

print('The "BM_venues" dataframe has {} venues and {} unique venue types.'.format(
      len(BM_venues['Venue Category']),
      len(BM_venues['Venue Category'].unique())))

BM_venues.to_csv('BM_venues.csv', sep=',', encoding='UTF8')
BM_venues.head()

Marble Hill
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker Heights
Gerritsen Beach
Marine Park
Clinton Hill
Sea Gate
Downtown
Boerum Hill
Prospect Lefferts Gardens
Ocean Hill
City Line
Bergen Beach
Midwood
Prospect Park South
Georgetown
East Williamsburg
North Side
South Side
Ocean Parkway
Fort Hamilton
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho

Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066    Bikram Yoga   
1  Marble Hill              40.876551               -73.91066       Arturo's   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066    Sam's Pizza   
4  Marble Hill              40.876551               -73.91066      Starbucks   

   Venue Latitude  Venue Longitude Venue Category  
0       40.876844       -73.906204    Yoga Studio  
1       40.874412       -73.910271    Pizza Place  
2       40.880404       -73.908937          Diner  
3       40.879435       -73.905859    Pizza Place  
4       40.877531       -73.905582    Coffee Shop

In [21]:
colnames = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
BM_venues = pd.read_csv('BM_venues.csv', skiprows=1, names=colnames)
BM_venues.columns = BM_venues.columns.str.replace(' ', '')
BM_venues.head()

Neighborhood  NeighborhoodLatitude  NeighborhoodLongitude          Venue  \
0  Marble Hill             40.876551              -73.91066    Bikram Yoga   
1  Marble Hill             40.876551              -73.91066       Arturo's   
2  Marble Hill             40.876551              -73.91066  Tibbett Diner   
3  Marble Hill             40.876551              -73.91066    Sam's Pizza   
4  Marble Hill             40.876551              -73.91066      Starbucks   

   VenueLatitude  VenueLongitude VenueCategory  
0      40.876844      -73.906204   Yoga Studio  
1      40.874412      -73.910271   Pizza Place  
2      40.880404      -73.908937         Diner  
3      40.879435      -73.905859   Pizza Place  
4      40.877531      -73.905582   Coffee Shop

In [22]:
BM_venues.shape

(9656, 7)

In [23]:
def Venues_Map(Borough_name, Borough_neighborhoods):
    
    # Use geopy library to get the latitude and longitude values 
    geolocator = Nominatim(user_agent="Jupyter")
    Borough_location = geolocator.geocode(Borough_name) #'Brooklyn, NY'
    Borough_latitude = Borough_location.latitude
    Borough_longitude = Borough_location.longitude
    print('The geographical coordinates of "{}" are {}, {}.'.format(Borough_name, Borough_latitude, Borough_longitude))
    
    # To verify the number of Boroughs and Neighborhoods in the extracted data
    print('The "{}" dataframe has {} different venue types and {} neighborhoods.'.format(
          Borough_name,
          len(Borough_neighborhoods['VenueCategory'].unique()),
          len(Borough_neighborhoods['Neighborhood'].unique())))
    
    # create map of city using latitude and longitude values
    map_Borough = folium.Map(location=[Borough_latitude, Borough_longitude], zoom_start=10)

    # add markers to map
    for lat, lng, venue, category in zip(Borough_neighborhoods['VenueLatitude'], Borough_neighborhoods['VenueLongitude'], Borough_neighborhoods['Venue'], Borough_neighborhoods['VenueCategory']):
        label = '{}, {}'.format(category, venue)
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=0.1,
            popup=label,
            color='red',
            fill=True,
            fill_color='#FF0000',
            fill_opacity=0.3).add_to(map_Borough)  

    return map_Borough

In [26]:
BM_venues.groupby('VenueCategory')['Venue'].count().sort_values(ascending=False)

VenueCategory
Pizza Place                                 411
Coffee Shop                                 364
Bakery                                      240
Park                                        235
Italian Restaurant                          229
Bar                                         216
Café                                        215
Grocery Store                               195
American Restaurant                         174
Mexican Restaurant                          157
Chinese Restaurant                          153
Sandwich Place                              151
Gym                                         149
Ice Cream Shop                              141
Deli / Bodega                               139
Cocktail Bar                                137
Caribbean Restaurant                        136
Gym / Fitness Center                        128
Wine Shop                                   126
Bagel Shop                                  117
Donut Shop                

In [27]:
BM_venues.groupby('Neighborhood').count()

NeighborhoodLatitude  NeighborhoodLongitude  Venue  \
Neighborhood                                                                    
Bath Beach                                   94                     94     94   
Battery Park City                           100                    100    100   
Bay Ridge                                   100                    100    100   
Bedford Stuyvesant                          100                    100    100   
Bensonhurst                                 100                    100    100   
Bergen Beach                                 10                     10     10   
Boerum Hill                                 100                    100    100   
Borough Park                                 55                     55     55   
Brighton Beach                               92                     92     92   
Broadway Junction                            58                     58     58   
Brooklyn Heights                            100                    100    100   
Brownsville                                  72                     72     72   
Bushwick                                    100                    100    100   
Canarsie                                     29                     29     29   
Carnegie Hill                               100                    100    100   
Carroll Gardens                             100                    100    100   
Central Harlem                              100                    100    100   
Chelsea                                     100                    100    100   
Chinatown                                   100                    100    100   
City Line                                    57                     57     57   
Civic Center                                100                    100    100   
Clinton                                     100                    100    100   
Clinton Hill                                100                    100    100   
Cobble Hill                                 100                    100    100   
Coney Island                                 98                     98     98   
Crown Heights                               100                    100    100   
Cypress Hills                                44                     44     44   
Ditmas Park                                 100                    100    100   
Downtown                                    100                    100    100   
Dumbo                                       100                    100    100   
Dyker Heights                                95                     95     95   
East Flatbush                                26                     26     26   
East Harlem                                 100                    100    100   
East New York                                27                     27     27   
East Village                                100                    100    100   
East Williamsburg                           100                    100    100   
Erasmus                                      93                     93     93   
Financial District                          100                    100    100   
Flatbush                                    100                    100    100   
Flatiron                                    100                    100    100   
Flatlands                                    60                     60     60   
Fort Greene                                 100                    100    100   
Fort Hamilton                               100                    100    100   
Fulton Ferry                                100                    100    100   
Georgetown                                   83                     83     83   
Gerritsen Beach                              65                     65     65   
Gowanus                                     100                    100    100   
Gramercy                                    100           

In [28]:
print('There are {} uniques categories.'.format(len(BM_venues['VenueCategory'].unique())))

There are 407 uniques categories.


### 3. Analyze Each Neighborhood

In [29]:
# one hot encoding
BM_onehot = pd.get_dummies(BM_venues[['VenueCategory']], prefix="", prefix_sep="")

#column lists before adding neighborhood
column_names = ['Neighborhood'] + list(BM_onehot.columns)

# add neighborhood column back to dataframe
BM_onehot['Neighborhood'] = BM_venues['Neighborhood'] 

# move neighborhood column to the first column
BM_onehot = BM_onehot[column_names]

BM_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  African Restaurant  \
0  Marble Hill                  0               0                   0   
1  Marble Hill                  0               0                   0   
2  Marble Hill                  0               0                   0   
3  Marble Hill                  0               0                   0   
4  Marble Hill                  0               0                   0   

   Airport Terminal  American Restaurant  Amphitheater  Animal Shelter  \
0                 0                    0             0               0   
1                 0                    0             0               0   
2                 0                    0             0               0   
3                 0                    0             0               0   
4                 0                    0             0               0   

   Antique Shop  Aquarium  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0             0         0       0                 0                       0   
1             0         0       0                 0                       0   
2             0         0       0                 0                       0   
3             0         0       0                 0                       0   
4             0         0       0                 0                       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0            0           0                    0                     0   
1            0           0                    0                     0   
2            0           0                    0                     0   
3            0           0                    0                     0   
4            0           0                    0                     0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0                 0                   0           0                      0   
1                 0                   0           0                      0   
2                 0                   0           0                      0   
3                 0                   0           0                      0   
4                 0                   0           0                      0   

   Austrian Restaurant  Auto Garage  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  \
0                    0            0          0           0       0     0    0   
1                    0            0          0           0       0     0    0   
2                    0            0          0           0       0     0    0   
3                    0            0          0           0       0     0    0   
4                    0            0          0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Bath House  Beach  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0           0      0                0         0            0           0   
1           0      0                0         0            0           0   
2           0      0                0         0            0           0   
3           0      0                0         0            0           0   
4           0      0                0         0            0           0   

   Big Box Store  Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  \
0              0                         0          0           0       0   
1              0                         0          0           0       0   
2              0                         0          0           0       0   
3              0    

In [30]:
restaurant_List = []
search = 'Restaurant'
for i in BM_onehot.columns :
    if search in i:
        restaurant_List.append(i)

In [31]:
restaurant_List

['African Restaurant',
 'American Restaurant',
 'Arepa Restaurant',
 'Argentinian Restaurant',
 'Asian Restaurant',
 'Australian Restaurant',
 'Austrian Restaurant',
 'Burmese Restaurant',
 'Cajun / Creole Restaurant',
 'Cambodian Restaurant',
 'Cantonese Restaurant',
 'Caribbean Restaurant',
 'Caucasian Restaurant',
 'Chinese Restaurant',
 'Colombian Restaurant',
 'Comfort Food Restaurant',
 'Cuban Restaurant',
 'Czech Restaurant',
 'Dim Sum Restaurant',
 'Dumpling Restaurant',
 'Eastern European Restaurant',
 'Empanada Restaurant',
 'English Restaurant',
 'Ethiopian Restaurant',
 'Falafel Restaurant',
 'Fast Food Restaurant',
 'Filipino Restaurant',
 'French Restaurant',
 'German Restaurant',
 'Greek Restaurant',
 'Halal Restaurant',
 'Hawaiian Restaurant',
 'Hotpot Restaurant',
 'Indian Restaurant',
 'Israeli Restaurant',
 'Italian Restaurant',
 'Japanese Curry Restaurant',
 'Japanese Restaurant',
 'Jewish Restaurant',
 'Korean Restaurant',
 'Kosher Restaurant',
 'Latin American Res

In [32]:
col_name = []
col_name = ['Neighborhood'] + restaurant_List
BM_restaurant = BM_onehot[col_name]
BM_restaurant = BM_restaurant.iloc[:,1::]

In [33]:
BM_restaurant_grouped = BM_restaurant.groupby('Neighborhood').sum().reset_index()

In [34]:
BM_restaurant_grouped['Total'] = BM_restaurant_grouped .sum(axis=1)

### 4. Cluster Neighborhoods and Examine Clusters

**_First, let's determine the optimal value of K for our dataset using the Silhouette Coefficient Method_**

In [35]:
BM_grouped_clustering = BM_restaurant_grouped.drop('Neighborhood', 1)

for n_cluster in range(2, 10):
    kmeans = KMeans(n_clusters=n_cluster).fit(BM_grouped_clustering)
    label = kmeans.labels_
    sil_coeff = silhouette_score(BM_grouped_clustering, label, metric='euclidean')
    print("For n_clusters={}, The Silhouette Coefficient is {}".format(n_cluster, sil_coeff))

For n_clusters=2, The Silhouette Coefficient is 0.3900812349669867
For n_clusters=3, The Silhouette Coefficient is 0.31826315229505725
For n_clusters=4, The Silhouette Coefficient is 0.2330677588086757
For n_clusters=5, The Silhouette Coefficient is 0.2330547642385833
For n_clusters=6, The Silhouette Coefficient is 0.18484698343934608
For n_clusters=7, The Silhouette Coefficient is 0.19547654482479038
For n_clusters=8, The Silhouette Coefficient is 0.1877446176219467
For n_clusters=9, The Silhouette Coefficient is 0.15771443818751027


**_As we can see, n_clusters=2 has highest Silhouette Coefficient. This means that 2 should be the optimal number of clusters.  
For n_clusters=2, The Silhouette Coefficient is 0.3900812349669867_**

In [36]:
# set number of clusters
kclusters = 2

BM_grouped_clustering = BM_restaurant_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(BM_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0],
      dtype=int32)

In [37]:
BM_results = pd.DataFrame(kmeans.cluster_centers_)
BM_results.columns = BM_grouped_clustering.columns
BM_results.index = ['cluster0','cluster1']
BM_results['Total Sum'] = BM_results.sum(axis = 1)
BM_results

African Restaurant  American Restaurant  Arepa Restaurant  \
cluster0            0.106061             1.787879      1.363636e-01   
cluster1            0.022727             1.272727     -6.938894e-17   

          Argentinian Restaurant  Asian Restaurant  Australian Restaurant  \
cluster0                0.090909          0.484848               0.136364   
cluster1                0.022727          0.068182               0.022727   

          Austrian Restaurant  Burmese Restaurant  Cajun / Creole Restaurant  \
cluster0             0.075758        1.515152e-02               4.545455e-02   
cluster1             0.022727       -1.734723e-18               3.469447e-18   

          Cambodian Restaurant  Cantonese Restaurant  Caribbean Restaurant  \
cluster0          4.545455e-02              0.090909              1.378788   
cluster1          3.469447e-18              0.022727              1.022727   

          Caucasian Restaurant  Chinese Restaurant  Colombian Restaurant  \
cluster0              0.045455            1.575758          1.515152e-02   
cluster1              0.045455            1.113636         -1.734723e-18   

          Comfort Food Restaurant  Cuban Restaurant  Czech Restaurant  \
cluster0             1.060606e-01          0.272727      3.030303e-02   
cluster1             1.387779e-17          0.090909     -3.469447e-18   

          Dim Sum Restaurant  Dumpling Restaurant  \
cluster0        9.090909e-02             0.121212   
cluster1        6.938894e-18             0.022727   

          Eastern European Restaurant  Empanada Restaurant  \
cluster0                     0.181818         1.060606e-01   
cluster1                     0.045455         1.387779e-17   

          English Restaurant  Ethiopian Restaurant  Falafel Restaurant  \
cluster0        4.545455e-02              0.106061            0.181818   
cluster1        3.469447e-18              0.022727            0.204545   

          Fast Food Restaurant  Filipino Restaurant  French Restaurant  \
cluster0              0.454545         1.212121e-01           0.606061   
cluster1              0.954545        -1.387779e-17           0.363636   

          German Restaurant  Greek Restaurant  Halal Restaurant  \
cluster0           0.090909          0.454545          0.121212   
cluster1           0.045455          0.045455          0.022727   

          Hawaiian Restaurant  Hotpot Restaurant  Indian Restaurant  \
cluster0         7.575758e-02       6.060606e-02           0.742424   
cluster1         2.775558e-17      -6.938894e-18           0.136364   

          Israeli Restaurant  Italian Restaurant  Japanese Curry Restaurant  \
cluster0        4.545455e-02            2.712121               3.030303e-02   
cluster1        3.469447e-18            1.136364              -3.469447e-18   

          Japanese Restaurant  Jewish Restaurant  Korean Restaurant  \
cluster0             1.166667       3.030303e-02           0.287879   
cluster1             0.545455      -3.469447e-18           0.045455   

          Kosher Restaurant  Latin American Restaurant  Lebanese Restaurant  \
cluster0           0.045455                   0.727273         1.060606e-01   
cluster1           0.045455                   0.295455         1.387779e-17   

          Malay Restaurant  Mediterranean Restaurant  Mexican Restaurant  \
cluster0      6.060606e-02                  0.469697            2.045455   
cluster1     -6.938894e-18                  0.295455            0.500000   

          Middle Eastern Restaurant  Molecular Gastronomy Restaurant  \
cluster0                   0.424242                         0.015152   
cluster1                   0.250000                         0.022727   

          Moroccan Restaurant  New American Restaurant  Paella Restaurant  \
cluster0         9.090909e-02                 0.666667       1.515152e-02   
cluster1         6.938894e-18                 0.181818      -1.734723e-18   

          Pakistani Restaurant  Persian Restaurant  Peruvian R

In [38]:
BM_results_merged = pd.DataFrame(BM_restaurant_grouped['Neighborhood'])

BM_results_merged['Total'] = BM_restaurant_grouped['Total']
BM_results_merged = BM_results_merged.assign(Cluster_Labels = kmeans.labels_)

In [39]:
print(BM_results_merged.shape)
BM_results_merged

(110, 3)


Neighborhood  Total  Cluster_Labels
0                   Bath Beach     30               0
1            Battery Park City     13               1
2                    Bay Ridge     36               0
3           Bedford Stuyvesant     20               0
4                  Bensonhurst     28               0
5                 Bergen Beach      5               1
6                  Boerum Hill     19               1
7                 Borough Park      9               1
8               Brighton Beach     21               0
9            Broadway Junction      9               1
10            Brooklyn Heights     18               1
11                 Brownsville     11               1
12                    Bushwick     25               0
13                    Canarsie      7               1
14               Carnegie Hill     22               0
15             Carroll Gardens     18               1
16              Central Harlem     36               0
17                     Chelsea     14               1
18                   Chinatown     35               0
19                   City Line     12               1
20                Civic Center     16               1
21                     Clinton     18               1
22                Clinton Hill     34               0
23                 Cobble Hill     20               0
24                Coney Island      7               1
25               Crown Heights     24               0
26               Cypress Hills     12               1
27                 Ditmas Park     33               0
28                    Downtown     19               1
29                       Dumbo     13               1
30               Dyker Heights     25               0
31               East Flatbush      6               1
32                 East Harlem     30               0
33               East New York      3               1
34                East Village     26               0
35           East Williamsburg     21               0
36                     Erasmus     20               0
37          Financial District     17               1
38                    Flatbush     32               0
39                    Flatiron     27               0
40                   Flatlands     11               1
41                 Fort Greene     28               0
42               Fort Hamilton     27               0
43                Fulton Ferry     13               1
44                  Georgetown     19               0
45             Gerritsen Beach     11               1
46                     Gowanus     20               0
47                    Gramercy     34               0
48                   Gravesend      9               1
49                  Greenpoint     23               0
50           Greenwich Village     32               0
51            Hamilton Heights     31               0
52               Highland Park      7               1
53                   Homecrest     25               0
54                Hudson Yards     13               1
55                      Inwood     34               0
56                  Kensington     22               0
57                  Lenox Hill     26               0
58              Lincoln Square     21               0
59                Little Italy     28               0
60             Lower East Side     23               0
61                     Madison     11               1
62             Manhattan Beach     23               0
63           Manhattan Terrace     17               1
64            Manhattan Valley     25               0
65              Manhattanville     33               0
66                 Marble Hill     14               1
67                 Marine Park     11               1
68                     Midtown     16               1
69               Midtown South     26               0
70                     Midwood      9               1
71                  Mill Basin     17               1
72                 Mill Island      3               1
73         Morningside

In [40]:
BM_merged = BM_Geo

BM_merged = BM_merged.join(BM_results_merged.set_index('Neighborhood'), on='Neighborhood')

print(BM_merged.shape)
BM_merged.head(10) # check the last columns!

(110, 6)


Borough       Neighborhood   Latitude  Longitude  Total  Cluster_Labels
0  Manhattan        Marble Hill  40.876551 -73.910660     14               1
1   Brooklyn          Bay Ridge  40.625801 -74.030621     36               0
2   Brooklyn        Bensonhurst  40.611009 -73.995180     28               0
3   Brooklyn        Sunset Park  40.645103 -74.010316     39               0
4   Brooklyn         Greenpoint  40.730201 -73.954241     23               0
5   Brooklyn          Gravesend  40.595260 -73.973471      9               1
6   Brooklyn     Brighton Beach  40.576825 -73.965094     21               0
7   Brooklyn     Sheepshead Bay  40.586890 -73.943186     31               0
8   Brooklyn  Manhattan Terrace  40.614433 -73.957438     17               1
9   Brooklyn           Flatbush  40.636326 -73.958401     32               0

In [41]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(BM_merged['Latitude'], BM_merged['Longitude'], BM_merged['Neighborhood'], BM_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## List Neighborhoods of Interest in New York City

### Cluster 0 : Saturated Markets

In [66]:
BM_merged[BM_merged['Cluster_Labels'] == 0].reset_index(drop=True)

Borough               Neighborhood   Latitude  Longitude  Total  \
0    Brooklyn                  Bay Ridge  40.625801 -74.030621     36   
1    Brooklyn                Bensonhurst  40.611009 -73.995180     28   
2    Brooklyn                Sunset Park  40.645103 -74.010316     39   
3    Brooklyn                 Greenpoint  40.730201 -73.954241     23   
4    Brooklyn             Brighton Beach  40.576825 -73.965094     21   
5    Brooklyn             Sheepshead Bay  40.586890 -73.943186     31   
6    Brooklyn                   Flatbush  40.636326 -73.958401     32   
7    Brooklyn              Crown Heights  40.670829 -73.943291     24   
8    Brooklyn                 Kensington  40.642382 -73.980421     22   
9    Brooklyn           Prospect Heights  40.676822 -73.964859     27   
10   Brooklyn               Williamsburg  40.707144 -73.958115     26   
11   Brooklyn                   Bushwick  40.698116 -73.925258     25   
12   Brooklyn         Bedford Stuyvesant  40.687232 -73.941785     20   
13   Brooklyn                Cobble Hill  40.687920 -73.998561     20   
14   Brooklyn                    Gowanus  40.673931 -73.994441     20   
15   Brooklyn                Fort Greene  40.688527 -73.972906     28   
16   Brooklyn                 Park Slope  40.672321 -73.977050     21   
17   Brooklyn            Manhattan Beach  40.577914 -73.943537     23   
18   Brooklyn                 Bath Beach  40.599519 -73.998752     30   
19   Brooklyn              Dyker Heights  40.619219 -74.019314     25   
20   Brooklyn               Clinton Hill  40.693229 -73.967843     34   
21   Brooklyn  Prospect Lefferts Gardens  40.658420 -73.954899     33   
22   Brooklyn        Prospect Park South  40.647009 -73.962613     34   
23   Brooklyn                 Georgetown  40.623845 -73.916075     19   
24   Brooklyn          East Williamsburg  40.708492 -73.938858     21   
25   Brooklyn                 North Side  40.714823 -73.958809     25   
26   Brooklyn                 South Side  40.710861 -73.958001     27   
27   Brooklyn              Ocean Parkway  40.613060 -73.968367     23   
28   Brooklyn              Fort Hamilton  40.614768 -74.031979     27   
29  Manhattan                  Chinatown  40.715618 -73.994279     35   
30  Manhattan         Washington Heights  40.851903 -73.936900     34   
31  Manhattan                     Inwood  40.867684 -73.921210     34   
32  Manhattan           Hamilton Heights  40.823604 -73.949688     31   
33  Manhattan             Manhattanville  40.816934 -73.957385     33   
34  Manhattan             Central Harlem  40.815976 -73.943211     36   
35  Manhattan                East Harlem  40.792249 -73.944182     30   
36  Manhattan            Upper East Side  40.775639 -73.960508     24   
37  Manhattan                  Yorkville  40.775930 -73.947118     26   
38  Manhattan                 Lenox Hill  40.768113 -73.958860     26   
39  Manhattan           Roosevelt Island  40.762160 -73.949168     24   
40  Manhattan            Upper West Side  40.787658 -73.977059     29   
41  Manhattan             Lincoln Square  40.773529 -73.985338     21   
42  Manhattan                Murray Hill  40.748303 -73.978332     25   
43  Manhattan          Greenwich Village  40.726933 -73.999914     32   
44  Manhattan               East Village  40.727847 -73.982226     26   
45  Manhattan            Lower East Side  40.717807 -73.980890     23   
46  Manhattan                    Tribeca  40.721522 -74.010683     22   
47  Manhattan               Little Italy  40.719324 -73.997305     28   
48  Manhattan                       Soho  40.722184 -74.000657     25   
49  Manhattan               West Village  40.734434 -74.006180     24   
50  Manhattan           Manhattan Valley  40.797307 -73.964286     25   
51  Manhattan        Morningside Heights  40.808000 -73.963896     33   
52  Manhattan                   Gramercy  40.737210 -73.981376     34   
53   Brooklyn                Ditmas Park  40.643675 -73.961013  

### Cluster 1 : Untapped Markets

In [67]:
BM_merged[BM_merged['Cluster_Labels'] == 1].reset_index(drop=True)

Borough        Neighborhood   Latitude  Longitude  Total  Cluster_Labels
0   Manhattan         Marble Hill  40.876551 -73.910660     14               1
1    Brooklyn           Gravesend  40.595260 -73.973471      9               1
2    Brooklyn   Manhattan Terrace  40.614433 -73.957438     17               1
3    Brooklyn       East Flatbush  40.641718 -73.936103      6               1
4    Brooklyn     Windsor Terrace  40.656946 -73.980073     18               1
5    Brooklyn         Brownsville  40.663950 -73.910235     11               1
6    Brooklyn    Brooklyn Heights  40.695864 -73.993782     18               1
7    Brooklyn     Carroll Gardens  40.680540 -73.994654     18               1
8    Brooklyn            Red Hook  40.676253 -74.012759     10               1
9    Brooklyn       Cypress Hills  40.682391 -73.876616     12               1
10   Brooklyn       East New York  40.669926 -73.880699      3               1
11   Brooklyn       Starrett City  40.647589 -73.879370      8               1
12   Brooklyn            Canarsie  40.635564 -73.902093      7               1
13   Brooklyn           Flatlands  40.630446 -73.929113     11               1
14   Brooklyn         Mill Island  40.606336 -73.908186      3               1
15   Brooklyn        Coney Island  40.574293 -73.988683      7               1
16   Brooklyn        Borough Park  40.633131 -73.990498      9               1
17   Brooklyn     Gerritsen Beach  40.590848 -73.930102     11               1
18   Brooklyn         Marine Park  40.609748 -73.931344     11               1
19   Brooklyn            Sea Gate  40.576375 -74.007873      1               1
20   Brooklyn            Downtown  40.690844 -73.983463     19               1
21   Brooklyn         Boerum Hill  40.685683 -73.983748     19               1
22   Brooklyn          Ocean Hill  40.678403 -73.913068     16               1
23   Brooklyn           City Line  40.678570 -73.867976     12               1
24   Brooklyn        Bergen Beach  40.615150 -73.898556      5               1
25   Brooklyn             Midwood  40.625596 -73.957595      9               1
26  Manhattan             Clinton  40.759101 -73.996119     18               1
27  Manhattan             Midtown  40.754691 -73.981669     16               1
28  Manhattan             Chelsea  40.744035 -74.003116     14               1
29  Manhattan   Battery Park City  40.711932 -74.016869     13               1
30  Manhattan  Financial District  40.707107 -74.010665     17               1
31   Brooklyn             Wingate  40.660947 -73.937187     13               1
32   Brooklyn               Rugby  40.655572 -73.926882     16               1
33  Manhattan        Civic Center  40.715229 -74.005415     16               1
34   Brooklyn            New Lots  40.662744 -73.885118      9               1
35   Brooklyn     Paerdegat Basin  40.631318 -73.902335      5               1
36   Brooklyn          Mill Basin  40.615974 -73.915154     17               1
37   Brooklyn        Fulton Ferry  40.703281 -73.995508     13               1
38   Brooklyn        Vinegar Hill  40.703321 -73.981116      9               1
39   Brooklyn   Broadway Junction  40.677861 -73.903317      9               1
40   Brooklyn               Dumbo  40.703176 -73.988753     13               1
41   Brooklyn       Highland Park  40.681999 -73.890346      7               1
42   Brooklyn             Madison  40.609378 -73.948415     11               1
43  Manhattan        Hudson Yards  40.756658 -74.000111     13               1

## Part 5B Segmenting and Clustering Neighborhoods - Bronx, Queens and Staten Island

## Introduction
In this section of the capstone project, we will use the Foursquare API to explore neighborhoods in Bronx, Queens and Staten Island.

## Table of Contents
1. Download and Explore Dataset
1. Explore Neighborhoods in Bronx, Queens and Staten Island
1. Analyze Each Neighborhood
1. Cluster Neighborhoods and Examine Clusters

### 1. Download and Explore Dataset

In [45]:
BQS_Geo = NYC_Geo.loc[(NYC_Geo['Borough'] == 'Bronx')|(NYC_Geo['Borough'] == 'Queens')|(NYC_Geo['Borough'] == 'Staten Island')]
BQS_Geo = BQS_Geo.reset_index(drop=True)
BQS_Geo.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [46]:
BQS_Geo.shape

(196, 4)

In [47]:
# create map using latitude and longitude values
map_BQS = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(BQS_Geo['Latitude'], BQS_Geo['Longitude'], BQS_Geo['Borough'], BQS_Geo['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_BQS)  
    
map_BQS

### 2. Explore Neighborhoods in Bronx, Queens and Staten Island

In [48]:
BQS_venues = getNearbyVenues(names=BQS_Geo['Neighborhood'],
                                  latitudes=BQS_Geo['Latitude'],
                                  longitudes=BQS_Geo['Longitude'],
                                  LIMIT=200)

print('The "BQS_venues" dataframe has {} venues and {} unique venue types.'.format(
      len(BQS_venues['Venue Category']),
      len(BQS_venues['Venue Category'].unique())))

BQS_venues.to_csv('BQS_venues.csv', sep=',', encoding='UTF8')
BQS_venues.head()

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Astoria
Woodside
Jackson Heights
Elmhurst
Howard Beach
Corona
Forest Hills
Kew Gardens
Richmond Hill
Flushing
Long Island City
Sunnyside
East Elmhurst
Maspeth
Ridgewood
Glendale
Rego Park
Woodhaven
Ozone Park
South Ozone Park
College Point
Whitestone
Bayside
Auburndale
Little Neck
Douglaston
Glen Oaks
Bellerose
Kew Gardens Hills
Fresh Meadows
Briarwood
Jamaica Center
Oakland Gardens
Queens Village
Hollis
South Jamaica
St. Albans
Rochdale
Springfield Gardens
Cam

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                         Venue  Venue Latitude  Venue Longitude  \
0             Lollipops Gelato       40.894123       -73.845892   
1           Ripe Kitchen & Bar       40.898152       -73.838875   
2              Ali's Roti Shop       40.894036       -73.856935   
3             Carvel Ice Cream       40.890487       -73.848568   
4  Jackie's West Indian Bakery       40.889283       -73.843310   

         Venue Category  
0          Dessert Shop  
1  Caribbean Restaurant  
2  Caribbean Restaurant  
3        Ice Cream Shop  
4  Caribbean Restaurant

In [49]:
colnames = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
BQS_venues = pd.read_csv('BQS_venues.csv', skiprows=1, names=colnames)
BQS_venues.columns = BQS_venues.columns.str.replace(' ', '')
BQS_venues.head()

Neighborhood  NeighborhoodLatitude  NeighborhoodLongitude  \
0    Wakefield             40.894705             -73.847201   
1    Wakefield             40.894705             -73.847201   
2    Wakefield             40.894705             -73.847201   
3    Wakefield             40.894705             -73.847201   
4    Wakefield             40.894705             -73.847201   

                         Venue  VenueLatitude  VenueLongitude  \
0             Lollipops Gelato      40.894123      -73.845892   
1           Ripe Kitchen & Bar      40.898152      -73.838875   
2              Ali's Roti Shop      40.894036      -73.856935   
3             Carvel Ice Cream      40.890487      -73.848568   
4  Jackie's West Indian Bakery      40.889283      -73.843310   

          VenueCategory  
0          Dessert Shop  
1  Caribbean Restaurant  
2  Caribbean Restaurant  
3        Ice Cream Shop  
4  Caribbean Restaurant

In [50]:
BQS_venues.groupby('VenueCategory')['Venue'].count().sort_values(ascending=False)

VenueCategory
Pizza Place                                 631
Donut Shop                                  343
Deli / Bodega                               325
Italian Restaurant                          298
Sandwich Place                              279
Pharmacy                                    279
Chinese Restaurant                          252
Bank                                        236
Bakery                                      228
Park                                        216
Coffee Shop                                 209
Fast Food Restaurant                        205
Grocery Store                               205
Bar                                         202
Ice Cream Shop                              183
Supermarket                                 180
Mexican Restaurant                          177
Bus Stop                                    170
Diner                                       169
Bus Station                                 137
American Restaurant       

In [51]:
BQS_venues.groupby('Neighborhood').count()

NeighborhoodLatitude  NeighborhoodLongitude  Venue  \
Neighborhood                                                               
Allerton                                66                     66     66   
Annadale                                19                     19     19   
Arden Heights                           22                     22     22   
Arlington                               22                     22     22   
Arrochar                                23                     23     23   
Arverne                                 34                     34     34   
Astoria                                100                    100    100   
Astoria Heights                         76                     76     76   
Auburndale                             100                    100    100   
Bay Terrace                             92                     92     92   
Baychester                             100                    100    100   
Bayside                                100                    100    100   
Bayswater                                7                      7      7   
Bedford Park                            75                     75     75   
Beechhurst                              44                     44     44   
Bellaire                                52                     52     52   
Belle Harbor                            26                     26     26   
Bellerose                               53                     53     53   
Belmont                                100                    100    100   
Blissville                              78                     78     78   
Bloomfield                              26                     26     26   
Breezy Point                             4                      4      4   
Briarwood                               73                     73     73   
Broad Channel                           16                     16     16   
Bronxdale                              100                    100    100   
Brookville                              18                     18     18   
Bulls Head                              57                     57     57   
Butler Manor                             8                      8      8   
Cambria Heights                         35                     35     35   
Castle Hill                             40                     40     40   
Castleton Corners                       50                     50     50   
Charleston                              56                     56     56   
Chelsea                                 39                     39     39   
City Island                             44                     44     44   
Claremont Village                       17                     17     17   
Clason Point                             7                      7      7   
Clifton                                 61                     61     61   
Co-op City                              81                     81     81   
College Point                           77                     77     77   
Concord                                 45                     45     45   
Concourse                               64                     64     64   
Concourse Village                      100                    100    100   
Corona                                 100                    100    100   
Country Club                            66                     66     66   
Dongan Hills                            62                     62     62   
Douglaston                              63                     63     63   
East Elmhurst                           86                     86     86   
East Tremont                            73                     73     73   
Eastchester                             71                     71     71   
Edenwald                                47                     47     47   
Edgemere                                23                     23     

In [52]:
print('There are {} uniques categories.'.format(len(BQS_venues['VenueCategory'].unique())))

There are 386 uniques categories.


### 3. Analyze Each Neighborhood

In [53]:
# one hot encoding
BQS_onehot = pd.get_dummies(BQS_venues[['VenueCategory']], prefix="", prefix_sep="")

#column lists before adding neighborhood
column_names = ['Neighborhood'] + list(BQS_onehot.columns)

# add neighborhood column back to dataframe
BQS_onehot['Neighborhood'] = BQS_venues['Neighborhood'] 

# move neighborhood column to the first column
BQS_onehot = BQS_onehot[column_names]

BQS_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  African Restaurant  \
0    Wakefield                  0                  0                   0   
1    Wakefield                  0                  0                   0   
2    Wakefield                  0                  0                   0   
3    Wakefield                  0                  0                   0   
4    Wakefield                  0                  0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0             0       0                 0                       0   
1             0       0                 0                       0   
2             0       0                 0                       0   
3             0       0                 0                       0   
4             0       0                 0                       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0            0           0                    0                     0   
1            0           0                    0                     0   
2            0           0                    0                     0   
3            0           0                    0                     0   
4            0           0                    0                     0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  \
0                 0                   0                      0   
1                 0                   0                      0   
2                 0                   0                      0   
3                 0                   0                      0   
4                 0                   0                      0   

   Auto Dealership  Auto Garage  Automotive Shop  BBQ Joint  Bagel Shop  \
0                0            0                0          0           0   
1                0            0                0          0           0   
2                0            0                0          0           0   
3                0            0                0          0           0   
4                0            0                0          0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0       0     0    0               0                 0                 0   
1       0     0    0               0                 0                 0   
2       0     0    0               0                 0                 0   
3       0     0    0               0                 0                 0   
4       0     0    0               0                 0                 0   

   Bath House  Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  \
0           0      0          0                0         0            0   
1           0      0          0                0         0            0   
2           0      0          0                0         0            0   
3           0      0          0                0         0            0   
4           0      0          0                0         0            0   

   Beer Store  Big Box Store  Bike Rental / Bike Share  Bike Trail  Bistro  \
0           0              0                         0           0       0   
1           0              0                         0           0       0   
2           0              0                         0           0       0   
3           0              0                         0           0       0   
4           0              0                         0           0       0   

   

In [54]:
restaurant_List1 = []
search = 'Restaurant'
for i in BQS_onehot.columns :
    if search in i:
        restaurant_List1.append(i)

In [55]:
restaurant_List1

['Afghan Restaurant',
 'African Restaurant',
 'American Restaurant',
 'Arepa Restaurant',
 'Argentinian Restaurant',
 'Asian Restaurant',
 'Australian Restaurant',
 'Brazilian Restaurant',
 'Cajun / Creole Restaurant',
 'Cantonese Restaurant',
 'Caribbean Restaurant',
 'Chinese Restaurant',
 'Colombian Restaurant',
 'Comfort Food Restaurant',
 'Cuban Restaurant',
 'Czech Restaurant',
 'Dim Sum Restaurant',
 'Dumpling Restaurant',
 'Eastern European Restaurant',
 'Egyptian Restaurant',
 'Empanada Restaurant',
 'Falafel Restaurant',
 'Fast Food Restaurant',
 'Filipino Restaurant',
 'French Restaurant',
 'German Restaurant',
 'Gluten-free Restaurant',
 'Greek Restaurant',
 'Halal Restaurant',
 'Himalayan Restaurant',
 'Hotpot Restaurant',
 'Indian Restaurant',
 'Indonesian Restaurant',
 'Italian Restaurant',
 'Japanese Restaurant',
 'Korean Restaurant',
 'Kosher Restaurant',
 'Latin American Restaurant',
 'Malay Restaurant',
 'Mediterranean Restaurant',
 'Mexican Restaurant',
 'Middle Eas

In [56]:
col_name = []
col_name = ['Neighborhood'] + restaurant_List1
BQS_restaurant = BQS_onehot[col_name]
BQS_restaurant = BQS_restaurant.iloc[:,1::]

In [57]:
BQS_restaurant_grouped = BQS_restaurant.groupby('Neighborhood').sum().reset_index()

In [58]:
BQS_restaurant_grouped['Total'] = BQS_restaurant_grouped .sum(axis=1)

### 4. Cluster Neighborhoods and Examine Clusters
**_First, let's determine the optimal value of K for our dataset using the Silhouette Coefficient Method_**

In [59]:
BQS_grouped_clustering = BQS_restaurant_grouped.drop('Neighborhood', 1)

for n_cluster in range(2, 10):
    kmeans = KMeans(n_clusters=n_cluster).fit(BQS_grouped_clustering)
    label = kmeans.labels_
    sil_coeff = silhouette_score(BQS_grouped_clustering, label, metric='euclidean')
    print("For n_clusters={}, The Silhouette Coefficient is {}".format(n_cluster, sil_coeff))

For n_clusters=2, The Silhouette Coefficient is 0.5188996477290404
For n_clusters=3, The Silhouette Coefficient is 0.3735381337383283
For n_clusters=4, The Silhouette Coefficient is 0.376145677539251
For n_clusters=5, The Silhouette Coefficient is 0.3746397526901166
For n_clusters=6, The Silhouette Coefficient is 0.33146717555328814
For n_clusters=7, The Silhouette Coefficient is 0.26418114856791125
For n_clusters=8, The Silhouette Coefficient is 0.258783603077586
For n_clusters=9, The Silhouette Coefficient is 0.26326650152524267


As we can see, n_clusters=2 has highest Silhouette Coefficient. This means that 2 should be the optimal number of clusters.
For n_clusters=2, The Silhouette Coefficient is 0.5188996477290404

In [60]:
# set number of clusters
kclusters = 2

BQS_grouped_clustering = BQS_restaurant_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(BQS_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0], dtype=int32)

In [61]:
BQS_results = pd.DataFrame(kmeans.cluster_centers_)
BQS_results.columns = BQS_grouped_clustering.columns
BQS_results.index = ['cluster0','cluster1']
BQS_results['Total Sum'] = BQS_results.sum(axis = 1)
BQS_results

Afghan Restaurant  African Restaurant  American Restaurant  \
cluster0           0.045455            0.045455             1.181818   
cluster1           0.006667            0.026667             0.566667   

          Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
cluster0          0.181818            2.500000e-01          0.886364   
cluster1          0.020000            1.387779e-16          0.253333   

          Australian Restaurant  Brazilian Restaurant  \
cluster0           2.272727e-02              0.181818   
cluster1           1.040834e-17              0.006667   

          Cajun / Creole Restaurant  Cantonese Restaurant  \
cluster0                   0.113636          1.590909e-01   
cluster1                   0.020000          6.245005e-17   

          Caribbean Restaurant  Chinese Restaurant  Colombian Restaurant  \
cluster0                  0.50            2.159091              0.113636   
cluster1                  0.56            1.046667              0.006667   

          Comfort Food Restaurant  Cuban Restaurant  Czech Restaurant  \
cluster0                 0.045455          0.090909      2.272727e-02   
cluster1                 0.013333          0.006667      1.040834e-17   

          Dim Sum Restaurant  Dumpling Restaurant  \
cluster0            0.159091             0.227273   
cluster1            0.013333             0.006667   

          Eastern European Restaurant  Egyptian Restaurant  \
cluster0                     0.181818         2.272727e-02   
cluster1                     0.060000         1.040834e-17   

          Empanada Restaurant  Falafel Restaurant  Fast Food Restaurant  \
cluster0             0.113636            0.022727              1.113636   
cluster1             0.013333            0.006667              1.040000   

          Filipino Restaurant  French Restaurant  German Restaurant  \
cluster0             0.204545           0.159091           0.022727   
cluster1             0.026667           0.013333           0.033333   

          Gluten-free Restaurant  Greek Restaurant  Halal Restaurant  \
cluster0           -3.469447e-18          0.795455          0.136364   
cluster1            6.666667e-03          0.060000          0.026667   

          Himalayan Restaurant  Hotpot Restaurant  Indian Restaurant  \
cluster0          6.818182e-02       1.136364e-01           1.454545   
cluster1          4.336809e-17       6.245005e-17           0.180000   

          Indonesian Restaurant  Italian Restaurant  Japanese Restaurant  \
cluster0           6.818182e-02            2.659091             0.795455   
cluster1           4.336809e-17            1.206667             0.253333   

          Korean Restaurant  Kosher Restaurant  Latin American Restaurant  \
cluster0           1.590909       4.545455e-02                       1.25   
cluster1           0.040000       2.081668e-17                       0.28   

          Malay Restaurant  Mediterranean Restaurant  Mexican Restaurant  \
cluster0      1.136364e-01                  0.204545            2.181818   
cluster1      6.245005e-17                  0.080000            0.540000   

          Middle Eastern Restaurant  Modern Greek Restaurant  \
cluster0                   0.386364             2.272727e-02   
cluster1                   0.093333             1.040834e-17   

          New American Restaurant  Paella Restaurant  Pakistani Restaurant  \
cluster0                 0.181818           0.022727              0.022727   
cluster1                 0.060000           0.013333              0.006667   

          Persian Restaurant  Peruvian Restaurant  Polish Restaurant  \
cluster0        2.272727e-02             0.477273           0.022727   
cluster1        1.040834e-17             0.066667           0.006667   

          Portuguese Restaurant  Ramen Restaurant  Restaurant  \
cluster0          -3.469447e-18      1.363636e-01    0.590909   
cluster1           6.666667e-03      8.673617e-17    0.340000   

          Romanian 

In [62]:
BQS_results_merged = pd.DataFrame(BQS_restaurant_grouped['Neighborhood'],)

BQS_results_merged['Total'] = BQS_restaurant_grouped['Total']
BQS_results_merged = BQS_results_merged.assign(Cluster_Labels = kmeans.labels_)

In [63]:
print(BQS_results_merged.shape)
BQS_results_merged

(194, 3)


Neighborhood  Total  Cluster_Labels
0                Allerton     11               1
1                Annadale      2               1
2           Arden Heights      3               1
3               Arlington      3               1
4                Arrochar      5               1
5                 Arverne      2               1
6                 Astoria     30               0
7         Astoria Heights     11               1
8              Auburndale     34               0
9             Bay Terrace     14               1
10             Baychester     15               1
11                Bayside     34               0
12              Bayswater      1               1
13           Bedford Park     10               1
14             Beechhurst     11               1
15               Bellaire      9               1
16           Belle Harbor      2               1
17              Bellerose      7               1
18                Belmont     26               0
19             Blissville     14               1
20             Bloomfield      0               1
21           Breezy Point      1               1
22              Briarwood     22               0
23          Broad Channel      1               1
24              Bronxdale     16               1
25             Brookville      4               1
26             Bulls Head     12               1
27           Butler Manor      0               1
28        Cambria Heights      8               1
29            Castle Hill      4               1
30      Castleton Corners      9               1
31             Charleston     10               1
32                Chelsea      3               1
33            City Island     15               1
34      Claremont Village      2               1
35           Clason Point      1               1
36                Clifton     14               1
37             Co-op City     13               1
38          College Point     11               1
39                Concord     11               1
40              Concourse     17               1
41      Concourse Village     20               0
42                 Corona     22               0
43           Country Club     14               1
44           Dongan Hills     20               0
45             Douglaston     22               0
46          East Elmhurst     18               0
47           East Tremont      8               1
48            Eastchester     12               1
49               Edenwald      9               1
50               Edgemere      2               1
51         Edgewater Park     12               1
52            Egbertville      2               1
53               Elm Park     17               1
54               Elmhurst     54               0
55            Eltingville     11               1
56           Emerson Hill      3               1
57           Far Rockaway      6               1
58              Fieldston      6               1
59            Floral Park     18               0
60               Flushing     44               0
61                Fordham     27               0
62           Forest Hills     33               0
63   Forest Hills Gardens     31               0
64              Fox Hills      5               1
65          Fresh Meadows     17               1
66              Glen Oaks     12               1
67               Glendale     10               1
68           Graniteville     13               1
69             Grant City     23               0
70               Grasmere     12               1
71            Great Kills     14               1
72             Greenridge      6               1
73            Grymes Hill      3               1
74                Hammels      6               1
75      Heartland Village      3               1
76           High  Bridge      5               1
77              Hillcrest     16               1
78                 Hollis      3               1
79             Holliswood     11               1
80           Howard Beach     18            

In [64]:
BQS_merged = BQS_Geo

BQS_merged = BQS_merged.join(BQS_results_merged.set_index('Neighborhood'), on='Neighborhood')

print(BQS_merged.shape)
BQS_merged.head(10) # check the last columns!

(196, 6)


Borough    Neighborhood   Latitude  Longitude  Total  Cluster_Labels
0   Bronx       Wakefield  40.894705 -73.847201      8               1
1   Bronx      Co-op City  40.874294 -73.829939     13               1
2   Bronx     Eastchester  40.887556 -73.827806     12               1
3   Bronx       Fieldston  40.895437 -73.905643      6               1
4   Bronx       Riverdale  40.890834 -73.912585      9               1
5   Bronx     Kingsbridge  40.881687 -73.902818     19               0
6   Bronx        Woodlawn  40.898273 -73.867315      6               1
7   Bronx         Norwood  40.877224 -73.879391     16               1
8   Bronx  Williamsbridge  40.881039 -73.857446     15               1
9   Bronx      Baychester  40.866858 -73.835798     15               1

In [65]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(BQS_merged['Latitude'], BQS_merged['Longitude'], BQS_merged['Neighborhood'], BQS_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### List Neighborhoods of Interest in New York City - Bronx, Queens and Staten Island

### Cluster 0 : Saturated Markets

In [70]:
BQS_merged[BQS_merged['Cluster_Labels'] == 0].reset_index(drop=True)

Borough          Neighborhood   Latitude  Longitude  Total  \
0           Bronx           Kingsbridge  40.881687 -73.902818     19   
1           Bronx               Fordham  40.860997 -73.896427     27   
2           Bronx               Melrose  40.819754 -73.909422     19   
3           Bronx           Parkchester  40.837938 -73.856003     24   
4           Bronx           Morris Park  40.847549 -73.850402     18   
5           Bronx               Belmont  40.857277 -73.888452     26   
6           Bronx             Unionport  40.829774 -73.850535     19   
7          Queens               Astoria  40.768509 -73.915654     30   
8          Queens              Woodside  40.746349 -73.901842     44   
9          Queens       Jackson Heights  40.751981 -73.882821     52   
10         Queens              Elmhurst  40.744049 -73.881656     54   
11         Queens          Howard Beach  40.654225 -73.838138     18   
12         Queens                Corona  40.742382 -73.856825     22   
13         Queens          Forest Hills  40.725264 -73.844475     33   
14         Queens         Richmond Hill  40.697947 -73.831833     19   
15         Queens              Flushing  40.764454 -73.831773     44   
16         Queens             Sunnyside  40.740176 -73.926916     32   
17         Queens         East Elmhurst  40.764073 -73.867041     18   
18         Queens             Ridgewood  40.708323 -73.901435     29   
19         Queens             Rego Park  40.728974 -73.857827     26   
20         Queens               Bayside  40.766041 -73.774274     34   
21         Queens            Auburndale  40.761730 -73.791762     34   
22         Queens           Little Neck  40.770826 -73.738898     24   
23         Queens            Douglaston  40.766846 -73.742498     22   
24         Queens             Briarwood  40.710935 -73.811748     22   
25         Queens        Jamaica Center  40.704657 -73.796902     26   
26         Queens              Steinway  40.775923 -73.902290     31   
27         Queens        Rockaway Beach  40.582802 -73.822361     19   
28         Queens           Murray Hill  40.764126 -73.812763     45   
29         Queens           Floral Park  40.741378 -73.708847     18   
30         Queens       Queensboro Hill  40.744572 -73.825809     19   
31         Queens            Ravenswood  40.761705 -73.931575     37   
32         Queens           Lefrak City  40.736075 -73.862525     25   
33         Queens          North Corona  40.754071 -73.857518     23   
34         Queens  Forest Hills Gardens  40.714611 -73.841022     31   
35  Staten Island              New Dorp  40.572572 -74.116479     22   
36  Staten Island             Sunnyside  40.612760 -74.097126     32   
37  Staten Island          Dongan Hills  40.588673 -74.096399     20   
38  Staten Island            Grant City  40.576216 -74.105856     23   
39         Queens         Jamaica Hills  40.711460 -73.796465     27   
40         Queens                Utopia  40.733500 -73.796717     21   
41          Bronx     Concourse Village  40.824780 -73.915847     20   
42         Queens         Hunters Point  40.743414 -73.953868     28   
43         Queens     Sunnyside Gardens  40.745652 -73.918193     38   
44         Queens          Queensbridge  40.756091 -73.945631     19   

    Cluster_Labels  
0                0  
1                0  
2                0  
3                0  
4                0  
5                0  
6                0  
7                0  
8                0  
9                0  
10               0  
11               0  
12               0  
13               0  
14               0  
15               0  
16               0  
17               0  
18               0  
19               0  
20               0  
21               0  
22               0  
23               0  
24               0  
25               0  
26               0  
27               0  
28               0  
29               0  
30               0  
31               0  
32  

### Cluster 1 : Untapped Markets

In [71]:
BQS_merged[BQS_merged['Cluster_Labels'] == 1].reset_index(drop=True)

Borough         Neighborhood   Latitude  Longitude  Total  \
0            Bronx            Wakefield  40.894705 -73.847201      8   
1            Bronx           Co-op City  40.874294 -73.829939     13   
2            Bronx          Eastchester  40.887556 -73.827806     12   
3            Bronx            Fieldston  40.895437 -73.905643      6   
4            Bronx            Riverdale  40.890834 -73.912585      9   
5            Bronx             Woodlawn  40.898273 -73.867315      6   
6            Bronx              Norwood  40.877224 -73.879391     16   
7            Bronx       Williamsbridge  40.881039 -73.857446     15   
8            Bronx           Baychester  40.866858 -73.835798     15   
9            Bronx       Pelham Parkway  40.857413 -73.854756     10   
10           Bronx          City Island  40.847247 -73.786488     15   
11           Bronx         Bedford Park  40.870185 -73.885512     10   
12           Bronx   University Heights  40.855727 -73.910416     13   
13           Bronx       Morris Heights  40.847898 -73.919672     12   
14           Bronx         East Tremont  40.842696 -73.887356      8   
15           Bronx           West Farms  40.839475 -73.877745      4   
16           Bronx         High  Bridge  40.836623 -73.926102      5   
17           Bronx           Mott Haven  40.806239 -73.916100     14   
18           Bronx          Port Morris  40.801664 -73.913221      3   
19           Bronx             Longwood  40.815099 -73.895788     16   
20           Bronx          Hunts Point  40.809730 -73.883315      4   
21           Bronx           Morrisania  40.823592 -73.901506     16   
22           Bronx            Soundview  40.821012 -73.865746      7   
23           Bronx         Clason Point  40.806551 -73.854144      1   
24           Bronx          Throgs Neck  40.815109 -73.816350     10   
25           Bronx         Country Club  40.844246 -73.824099     14   
26           Bronx   Westchester Square  40.840619 -73.842194     16   
27           Bronx             Van Nest  40.843608 -73.866299     12   
28           Bronx       Spuyten Duyvil  40.881395 -73.917190     11   
29           Bronx      North Riverdale  40.908543 -73.904531     10   
30           Bronx           Pelham Bay  40.850641 -73.832074     14   
31           Bronx        Schuylerville  40.826580 -73.826203     11   
32           Bronx       Edgewater Park  40.821986 -73.813885     12   
33           Bronx          Castle Hill  40.819014 -73.848027      4   
34           Bronx            Olinville  40.871371 -73.863324     13   
35           Bronx       Pelham Gardens  40.862966 -73.841612      5   
36           Bronx            Concourse  40.834284 -73.915589     17   
37           Bronx             Edenwald  40.884561 -73.848083      9   
38          Queens          Kew Gardens  40.705179 -73.829819     11   
39          Queens     Long Island City  40.750217 -73.939202     16   
40          Queens              Maspeth  40.725427 -73.896217      9   
41          Queens             Glendale  40.702762 -73.870742     10   
42          Queens            Woodhaven  40.689887 -73.858110     15   
43          Queens           Ozone Park  40.680708 -73.843203     13   
44          Queens     South Ozone Park  40.668550 -73.809865      9   
45          Queens        College Point  40.784903 -73.843045     11   
46          Queens           Whitestone  40.781291 -73.814202     13   
47          Queens            Glen Oaks  40.749441 -73.715481     12   
48          Queens            Bellerose  40.728573 -73.720128      7   
49          Queens    Kew Gardens Hills  40.722578 -73.820878     13   
50          Queens        Fresh Meadows  40.734394 -73.782713     17   
51          Queens      Oakland Gardens  40.745619 -73.754950     14   
52          Queens       Queens Village  40.718893 -73.738715      7   
53          Queens               Hollis  40.711243 -73.759250      3   
54          Queens        South Jamaica  40.696911 